### PR TITLE
Adding OCP/k8s cluster ID to driver container environment variable.

### PR DESCRIPTION
**Pull request by nitishkumar4**
## Pull request checklist

- Fixes https://github.com/IBM/ibm-spectrum-scale-csi/issues/641

## Pull request type

 

 

Please check the type of change your PR introduces:
- [x] Bugfix
- [x] Feature Enhancement
- [ ] Test Automation 
- [ ] Code Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Community Operator listing 
- [ ] Other (please describe): 

## What is the current behavior?

-

## What is the new behavior?

- CGPrefix spec field is added to CR manifest.
- CGPrefix is optional
- User should pass a UUID string for the cluster in use, as value for `consistencyGroupPrefix`.
- If No value is passed.
   - For OCP, OCP cluster ID is used.
   - For Vanilla k8s, UUID is generated and used.
- Value of CGPrefix is stored as a environment variable `CSI_CG_PREFIX` in driver container.
- spec field CGPrefix is updated with the UUID.

## How risky is this change?
- [ ] Small, isolated change
- [x] Medium, requires regression testing
- [ ] Large, requires functional and regression testing

### CODE REVIEW

**Comment by nitishkumar4** [comment]
```sh 
# oc rsh ibm-spectrum-scale-csi-2fkgf env | grep CSI_CG_PREFIX
CSI_CG_PREFIX=bca0bb7b-c181-4ca7-9db2-e49f4af657a5
```

**Comment by nitishkumar4** [comment]
```sh
oc describe cso -n ibm-spectrum-scale-csi-driver
...
Spec:
  Consistency Group Prefix:  605855cb-ed40-4a7e-80a5-3134958d272d
...
```

**Comment by nitishkumar4** [comment]
LOGS

Kubernetes
```sh
1.6460387818051775e+09  INFO    csiscaleoperator_controller.GetConsistencyGroupPrefix   Checking if consistency group prefix is passed in CSIScaleOperator specs.
1.646038781805183e+09   INFO    csiscaleoperator_controller.GetConsistencyGroupPrefix   Consistency group prefix is not found in CSIScaleOperator specs.
1.646038781805188e+09   INFO    csiscaleoperator_controller.GetConsistencyGroupPrefix   Fetching cluster information.
1.6460387818051941e+09  INFO    csiscaleoperator_controller.GetConsistencyGroupPrefix   Cluster is a Kubernetes Platform.
1.6460387818051996e+09  INFO    csiscaleoperator_controller.GenerateUUID        Generating a unique cluster ID.
1.646038781805216e+09   INFO    csiscaleoperator_controller.Reconcile   Updating consistency group prefix in CSIScaleOperator resource.
1.6460387818171582e+09  INFO    csiscaleoperator_controller.Reconcile   Successfully updated consistency group prefix in CSIScaleOperator resource.
```

OCP:
```sh
1.6460316972816656e+09  INFO    csiscaleoperator_controller.GetConsistencyGroupPrefix   Checking if consistency group prefix is passed in CSIScaleOperator specs.
1.646031697281668e+09   INFO    csiscaleoperator_controller.GetConsistencyGroupPrefix   Consistency group prefix is not found in CSIScaleOperator specs.
1.64603169728167e+09    INFO    csiscaleoperator_controller.GetConsistencyGroupPrefix   Fetching cluster information.
1.6460316972816734e+09  INFO    csiscaleoperator_controller.GetConsistencyGroupPrefix   Cluster is Redhat Openshift Cluster Platform.
1.6460316972816775e+09  INFO    csiscaleoperator_controller.GetConsistencyGroupPrefix   Fetching cluster ID from ClusterVersion resource.
1.6460316973828335e+09  INFO    csiscaleoperator_controller.Reconcile   Updating consistency group prefix in CSIScaleOperator resource.
1.6460316973950925e+09  INFO    csiscaleoperator_controller.Reconcile   Successfully updated consistency group prefix in CSIScaleOperator resource.

```

**Review by amdabhad — APPROVED** [review]
LGTM

File: driver/csiplugin/connectors/connectors.go [hunk 1]
```diff
@@ -108,6 +108,7 @@ const (
 	UserSpecifiedCompression      string = "compression"
 	UserSpecifiedTier             string = "tier"
 	UserSpecifiedSnapWindow       string = "snapWindow"
```
Comment (Jainbrt) [inline driver/csiplugin/connectors/connectors.go:110]: snapWindow is volumesnapshotclass parameter, not storageclass. Is there any specific reason it is here. Sorry though this change is not related to the PR.

File: operator/config/rbac/role.yaml [hunk 2]
```diff
@@ -39,6 +39,12 @@ rules:
   verbs:
   - get
   - update
+- apiGroups:
+  - config.openshift.io
+  resources:
+  - clusterversions
+  verbs:
+  - '*'
```
Comment (Jainbrt) [inline operator/config/rbac/role.yaml:47]: I think only get verb should be sufficient here ?

File: driver/csiplugin/connectors/connectors.go (hunk 1 above)
Comment (amdabhad) [reply driver/csiplugin/connectors/connectors.go:110]: It is just a constant defined here along with other input parameter keys, snapWindow is used for create snapshot only.

**Review by madhuthorat — APPROVED** [review]
LGTM. Good work !

File: operator/config/rbac/role.yaml (hunk 2 above)
Comment (madhuthorat) [reply operator/config/rbac/role.yaml:47]: We can handle this comment separately, at all other places for verbs we have "*"

Comment (Jainbrt) [reply operator/config/rbac/role.yaml:47]: @madhuthorat @nitishkumar4 did we open the issue/PR for the same ?

Comment (madhuthorat) [reply operator/config/rbac/role.yaml:47]: Being tracked at https://github.com/IBM/ibm-spectrum-scale-csi/issues/662
